### PR TITLE
Improve handling of indexed array access in flow analysis

### LINKED ISSUE #6616: CA1508 False Positive after array values set in a for loop
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/CA1508-False-Positive-after-array-values/10351632)._

---
[severity:It bothers me. A fix would be nice]
The attached code is a very simplified example of declaring an array of objects and then populating them in a for loop.
On exit, a test is done to ensure all objects in the array are not null.
Reality is a larger array and more complex population, but the attachment demonstrates the problem with a single array element.
Setting .Net analysis level to "Latest All" is required; the others "Latest" etc. don't show this warning.
Adding pragmas suppresses the warning.

When the object(s) are populated in the for loop, the warning is that the test for "is null" is always true.
More reasonable would be that it is unknown except at runtime.

p.s.
All this null checking is tremendously complex and I appreciate the ton of work that's been put into this tool and the (generally) fine job done.

---
#### green4don on 4/29/2023, 00:24 AM: 

(private comment, text removed)
#### Feedback Bot on 4/30/2023, 08:05 AM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #6616

We now ensure that setting an abstract value for an indexed access entity with any non-constant index clears out the existing value for entities that can overlap the indexed location

### CODE REVIEW

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/AnalysisEntityBasedPredicateAnalysisData.cs [hunk 1]
```diff
@@ -159,6 +163,46 @@ public PredicateValueKind ApplyPredicatedDataForEntity(AnalysisEntity predicated
 
         public void AddTrackedEntities(HashSet<AnalysisEntity> builder) => builder.UnionWith(CoreAnalysisData.Keys);
 
+        private void ClearOverlappingAnalysisDataForIndexedEntity(AnalysisEntity analysisEntity, TValue value)
```
Comment (mavasani) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/AnalysisEntityBasedPredicateAnalysisData.cs:166]: This is the core fix

**Comment by mavasani** [comment]
@Youssef1313

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/AnalysisEntity.cs [hunk 2]
```diff
@@ -204,6 +204,30 @@ or PointsToAbstractValueKind.UnknownNull
...
+                return true;
+            }
+
+            if (other == null ||
+                EqualsIgnoringInstanceLocationId != other.EqualsIgnoringInstanceLocationId)
+            {
+                return false;
+            }
+
+            // Now perform slow check that compares individual hash code parts sequences.
```
Comment (Youssef1313) [inline src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/AnalysisEntity.cs:221]: Why are we comparing hash codes?

**Review by Youssef1313 — APPROVED** [review]

File: src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/AnalysisEntity.cs (hunk 2 above)
Comment (mavasani) [reply src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/AnalysisEntity.cs:221]: @Youssef1313 We were seeing extremely large allocations and CPU stacks for dictionary operations keyed on `AnalysisEntity` during flow analysis, hence we had switched to using a CacheBasedEquatable approach. It's possible that this is no longer required in the current state of the flow analysis framework, but we would need to invest a few cycles in cleanup and performance measurements on real world code to confirm.

**Comment by mavasani** [comment]
@CollinAlpert I am seeing intermittent test failures for CA1001 in CI after https://github.com/dotnet/roslyn-analyzers/pull/6741

Test: 
`Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines.UnitTests.TypesThatOwnDisposableFieldsShouldBeDisposableAnalyzerTests.CA1001_ExcludedSymbolNamesAsync(editorConfigText: "")`

```
Expected diagnostic message to be "Type 'SomeClass' owns disposable field(s) '_fs', '_ms' but is not disposable" was "Type 'SomeClass' owns disposable field(s) '_fs' but is not disposable"


Expected diagnostic:
// /0/Test0.cs(4,14,4,23): hidden CA1001: Type 'SomeClass' owns disposable field(s) '_fs', '_ms' but is not disposable
VerifyCS.Diagnostic().WithSpan(4, 14, 4, 23).WithArguments("SomeClass", "_fs', '_ms"),


Actual diagnostic:
// /0/Test0.cs(4,14): hidden CA1001: Type 'SomeClass' owns disposable field(s) '_fs' but is not disposable
VerifyCS.Diagnostic().WithSpan(4, 14, 4, 23).WithArguments("SomeClass", "_fs"),


Assert.Equal() Failure
Expected: Type 'SomeClass' owns disposable field(s) '_fs', '_ms' but is not disposable
Actual:   Type 'SomeClass' owns disposable field(s) '_fs' but is not disposable



Stack trace
   at Microsoft.CodeAnalysis.Testing.Verifiers.XUnitVerifier.Equal[T](T expected, T actual, String message) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Testing.Verifiers.XUnit/XUnitVerifier.cs:line 49
   at Microsoft.CodeAnalysis.Testing.AnalyzerTest`1.VerifyDiagnosticResults(IEnumerable`1 actualResults, ImmutableArray`1 analyzers, DiagnosticResult[] expectedResults, IVerifier verifier) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs:line 603
   at Microsoft.CodeAnalysis.Testing.AnalyzerTest`1.VerifyDiagnosticsAsync(EvaluatedProjectState primaryProject, ImmutableArray`1 additionalProjects, DiagnosticResult[] expected, IVerifier verifier, CancellationToken cancellationToken) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs:line 465
   at Microsoft.CodeAnalysis.Testing.CodeFixTest`1.RunImplAsync(CancellationToken cancellationToken) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.CodeFix.Testing/CodeFixTest`1.cs:line 305
   at Microsoft.CodeAnalysis.Testing.AnalyzerTest`1.RunAsync(CancellationToken cancellationToken) in /_/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs:line 188
   at Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines.UnitTests.TypesThatOwnDisposableFieldsShouldBeDisposableAnalyzerTests.CA1001_ExcludedSymbolNamesAsync(String editorConfigText) in /_/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/TypesThatOwnDisposableFieldsShouldBeDisposableTests.cs:line 875
--- End of stack trace from previous location ---
```

**Pull request merged**